torch/testing/_internal/common_utils.py

```diff
@@ -777,27 +777,67 @@ def safeCoalesce(self, t):
 
         return tg
 
-    def assertEqual(self, x, y, prec=None, message='', allow_inf=False, exact_dtype=None):
+    # Some analysis of tolerance by logging tests from test_torch.py can be found
+    # in https://github.com/pytorch/pytorch/pull/32538.
+    # dtype name : (rtol, atol)
+    dtype_precisions = {
+        'float16': (0.001, 1e-5),
+        'bfloat16': (0.016, 1e-5),
+        'float32': (1.3e-6, 1.3e-6),
+        'float64': (1e-5, 1e-8),
+        'complex32': (1.3e-6, 1.3e-6),
+        'complex64': (1e-5, 1e-8),
+    }
+
+    def get_default_tolerance(self, a, b=None):
+        if b is None:
+            dtype = torch.float
+            if isinstance(a, torch.Tensor):
+                dtype = a.dtype
+            elif isinstance(a, torch.dtype):
+                dtype = a
+            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
+                dtype = torch.from_numpy(a).dtype
+
+            # todo: set 0-tolerance for integral type.
```

Comment:
To do this properly I don't want to have lists of dtypes, and we don't have an `is_integral` in python (`not is_floating_point` will include quantized types). I have a PR here to add is_integral:
https://github.com/pytorch/pytorch/pull/34312
but there are some outstanding questions on the approach, so I'd rather finish this and tighten integral tolerances in a subsequent PR.